db/mongo/client.go

```diff
@@ -0,0 +1,136 @@
+package mongo
+
+import (
+	"context"
+	"fmt"
+
+	wraperrors "github.com/pkg/errors"
+	guuid "github.com/satori/go.uuid"
+	"github.com/spf13/viper"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github/wasimkhan042/uStore-Auth/config"
+	"github/wasimkhan042/uStore-Auth/db"
+	"github/wasimkhan042/uStore-Auth/models"
+)
+
+const (
+	userCollection  = "user"
+	subscriptionCollection = "subscription"
+)
+
+func init() {
+	db.Register("mongo", NewMongoClient)
+}
+
+type client struct {
+	conn *mongo.Client
+}
+
+// NewMongoClient initializes a mongo database connection.
+func NewMongoClient(conf db.Option) (db.DataStore, error) {
+	uri := fmt.Sprintf("mongodb://%s:%s", viper.GetString(config.DbHost), viper.GetString(config.DbPort))
+	log().Infof("initializing mongodb: %s", uri)
+
+	cli, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
+	if err != nil {
+		return nil, wraperrors.Wrap(err, "failed to connect")
+	}
+
+	return &client{conn: cli}, nil
+}
+
+// Registration stores new user details in 'user' collection
+func (m client) Registration(userInfo *models.User) error {
+	_, err := m.ViewProfile(userInfo.Email)
+	// add unregistered users only
+	if err != nil {
+		if userInfo.ID != "" {
+			return nil
+		}
+		userInfo.ID = guuid.NewV4().String()
+		collection := m.conn.Database(viper.GetString(config.DbName)).Collection(userCollection)
+		if _, err := collection.InsertOne(context.TODO(), userInfo); err != nil {
+			return wraperrors.Wrap(err, "failed to add user")
+		}
+
+		return nil
+	}
+
+	return err
```

Comment:
do we need to return err here?

Reply:
err is returned (if the user is already registered) to service layer to manage logic. Else will return nil if the user successfully get register